File: Data Structures & Algorithms/binary-tree-diameter/submission-3.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number}
     */
    diameterOfBinaryTree(root) {
        let result = 0;  // lives here, persists across all recursive calls

        const dfs = (node) => {
            if (!node) return 0;

            let left = dfs(node.left);
            let right = dfs(node.right);

            result = Math.max(result, left + right);  // updates the outer result

            return 1 + Math.max(left, right);
        }

        dfs(root);
        return result;
    }
}

File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def lowestCommonAncestor(self, root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:
        # if not root: 
        #     return None

        # if root.val == p or root.val == q: 
        #     return root

        # left_result = self.lowestCommonAncestor(root.left, p, q)
        # right_result = self.lowestCommonAncestor(root.right, p, q)

        # if left_result and right_result: 
        #     return root 
        
        # return left_result if left_result else right_result
        if root.val > max(p.val, q.val):
            return self.lowestCommonAncestor(root.left, p, q)
        elif root.val < min(p.val, q.val):
            return self.lowestCommonAncestor(root.right, p, q)
        else:
            return root